UIA/SearchField: announce how many suggestions are available when entering text. re #7330

josephsl opened the pull request:
Hi,

Resolves the following:

* #7330: announces suggestions count when entering text in places such as Windows 10 Settings app and others.

Suggested what's new entry:

Section: New features

* NVDA can now announce how many suggestions are available when entering text in edit fields such as search box in Windows 10 Settings app.

Thanks.

josephsl committed:
NVDAObjects/UIA/SearchField: announce suggestions count for search boxes. re #7330.

In some search fields, because one needs to use arrow keys to go through suggestions, it is important that people know how many suggestions are out there. Thus suggestionsOpened event in SearchField will announce this provided that report position information checkbox is checked in Object Presentation dialog.

josephsl committed:
User guide: document another use of object position info checkbox in Object Presentation dialog.

derekriemer commented:
I noticed, although it could be me or something, that this caused a perf
regression, and disabling win10apps made it better. Not sure if there's any
reality to it.

On Tue, Jun 27, 2017 at 7:35 PM, Joseph Lee <notifications@github.com>
wrote:

> Hi,
>
> Resolves the following:
>
>    - #7330 <https://github.com/nvaccess/nvda/issues/7330>: announces
>    suggestions count when entering text in places such as Windows 10 Settings
>    app and others.
>
> Suggested what's new entry:
>
> Section: New features
>
>    - NVDA can now announce how many suggestions are available when
>    entering text in edit fields such as search box in Windows 10 Settings app.
>
> Thanks.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/nvaccess/nvda/pull/7331
> Commit Summary
>
>    - NVDAObjects/UIA/SearchField: announce suggestions count for search
>    boxes. re #7330.
>    - User guide: document another use of object position info checkbox in
>    Object Presentation dialog.
>
> File Changes
>
>    - *M* source/NVDAObjects/UIA/__init__.py
>    <https://github.com/nvaccess/nvda/pull/7331/files#diff-0> (13)
>    - *M* user_docs/en/userGuide.t2t
>    <https://github.com/nvaccess/nvda/pull/7331/files#diff-1> (1)
>
> Patch Links:
>
>    - https://github.com/nvaccess/nvda/pull/7331.patch
>    - https://github.com/nvaccess/nvda/pull/7331.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7331>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFGivRfaFzt5IHywUg-nnm28vsYyEVMlks5sIa3UgaJpZM4OHYuZ>
> .
>



-- 

Derek Riemer: Improving the world one byte at a time!

   - University of Colorado Boulder Department of computer science, 4th
   year undergraduate student.
   - Accessibility enthusiast.
   - Proud user of the NVDA screen reader.
   - Open source enthusiast.
   - Skier.

Personal website <http://derekriemer.com>

josephsl commented:
Eh? In that case, it isn’t worth it to proceed with this PR unless we find a much better way. Thanks.

 

From: Derek Riemer [mailto:notifications@github.com] 
Sent: Tuesday, June 27, 2017 8:01 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA/SearchField: announce how many suggestions are available when entering text. re #7330 (#7331)

 

I noticed, although it could be me or something, that this caused a perf
regression, and disabling win10apps made it better. Not sure if there's any
reality to it.

On Tue, Jun 27, 2017 at 7:35 PM, Joseph Lee <notifications@github.com>
wrote:

> Hi,
>
> Resolves the following:
>
> - #7330 <https://github.com/nvaccess/nvda/issues/7330>: announces
> suggestions count when entering text in places such as Windows 10 Settings
> app and others.
>
> Suggested what's new entry:
>
> Section: New features
>
> - NVDA can now announce how many suggestions are available when
> entering text in edit fields such as search box in Windows 10 Settings app.
>
> Thanks.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
> https://github.com/nvaccess/nvda/pull/7331
> Commit Summary
>
> - NVDAObjects/UIA/SearchField: announce suggestions count for search
> boxes. re #7330.
> - User guide: document another use of object position info checkbox in
> Object Presentation dialog.
>
> File Changes
>
> - *M* source/NVDAObjects/UIA/__init__.py
> <https://github.com/nvaccess/nvda/pull/7331/files#diff-0> (13)
> - *M* user_docs/en/userGuide.t2t
> <https://github.com/nvaccess/nvda/pull/7331/files#diff-1> (1)
>
> Patch Links:
>
> - https://github.com/nvaccess/nvda/pull/7331.patch
> - https://github.com/nvaccess/nvda/pull/7331.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7331>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFGivRfaFzt5IHywUg-nnm28vsYyEVMlks5sIa3UgaJpZM4OHYuZ>
> .
>



-- 

Derek Riemer: Improving the world one byte at a time!

- University of Colorado Boulder Department of computer science, 4th
year undergraduate student.
- Accessibility enthusiast.
- Proud user of the NVDA screen reader.
- Open source enthusiast.
- Skier.

Personal website <http://derekriemer.com>



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/7331#issuecomment-311543549> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkAPPXcTHpivFtJqlRJXVkyItWPCGks5sIcHSgaJpZM4OHYuZ> .

derekriemer commented:
I don't know if the effect is real, take it as a user being suspicious, but
not certain.

On Tue, Jun 27, 2017 at 9:05 PM, Joseph Lee <notifications@github.com>
wrote:

> Eh? In that case, it isn’t worth it to proceed with this PR unless we find
> a much better way. Thanks.
>
>
>
> From: Derek Riemer [mailto:notifications@github.com]
> Sent: Tuesday, June 27, 2017 8:01 PM
> To: nvaccess/nvda <nvda@noreply.github.com>
> Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <
> author@noreply.github.com>
> Subject: Re: [nvaccess/nvda] UIA/SearchField: announce how many
> suggestions are available when entering text. re #7330 (#7331)
>
>
>
> I noticed, although it could be me or something, that this caused a perf
> regression, and disabling win10apps made it better. Not sure if there's any
> reality to it.
>
> On Tue, Jun 27, 2017 at 7:35 PM, Joseph Lee <notifications@github.com>
> wrote:
>
> > Hi,
> >
> > Resolves the following:
> >
> > - #7330 <https://github.com/nvaccess/nvda/issues/7330>: announces
> > suggestions count when entering text in places such as Windows 10
> Settings
> > app and others.
> >
> > Suggested what's new entry:
> >
> > Section: New features
> >
> > - NVDA can now announce how many suggestions are available when
> > entering text in edit fields such as search box in Windows 10 Settings
> app.
> >
> > Thanks.
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> > https://github.com/nvaccess/nvda/pull/7331
> > Commit Summary
> >
> > - NVDAObjects/UIA/SearchField: announce suggestions count for search
> > boxes. re #7330.
> > - User guide: document another use of object position info checkbox in
> > Object Presentation dialog.
> >
> > File Changes
> >
> > - *M* source/NVDAObjects/UIA/__init__.py
> > <https://github.com/nvaccess/nvda/pull/7331/files#diff-0> (13)
> > - *M* user_docs/en/userGuide.t2t
> > <https://github.com/nvaccess/nvda/pull/7331/files#diff-1> (1)
> >
> > Patch Links:
> >
> > - https://github.com/nvaccess/nvda/pull/7331.patch
> > - https://github.com/nvaccess/nvda/pull/7331.diff
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/nvaccess/nvda/pull/7331>, or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AFGivRfaFzt5IHywUg-
> nnm28vsYyEVMlks5sIa3UgaJpZM4OHYuZ>
> > .
> >
>
>
>
> --
>
> Derek Riemer: Improving the world one byte at a time!
>
> - University of Colorado Boulder Department of computer science, 4th
> year undergraduate student.
> - Accessibility enthusiast.
> - Proud user of the NVDA screen reader.
> - Open source enthusiast.
> - Skier.
>
> Personal website <http://derekriemer.com>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <
> https://github.com/nvaccess/nvda/pull/7331#issuecomment-311543549> , or
> mute the thread <https://github.com/notifications/unsubscribe-auth/
> AHgLkAPPXcTHpivFtJqlRJXVkyItWPCGks5sIcHSgaJpZM4OHYuZ> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7331#issuecomment-311544158>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFGivQ1xRxkIIrHHXBS_gT477DELOMpdks5sIcL7gaJpZM4OHYuZ>
> .
>



-- 

Derek Riemer: Improving the world one byte at a time!

   - University of Colorado Boulder Department of computer science, 4th
   year undergraduate student.
   - Accessibility enthusiast.
   - Proud user of the NVDA screen reader.
   - Open source enthusiast.
   - Skier.

Personal website <http://derekriemer.com>

dkager commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1465,6 +1467,17 @@ def event_UIA_controllerFor(self):
+			if self.UIAElement.cachedAutomationID == "TextBox" or self.UIAElement.cachedAutomationID == "SearchTextBox" and self.appModule.appName != "searchui":
+				# Item count must be the last one announced.
+				# Translators: presented when there are suggestions as theu ser types (example output: 3 suggestions).
+				queueHandler.queueFunction(queueHandler.eventQueue, ui.message, _("{suggestionsCount} suggestions").format(suggestionsCount = self.controllerFor[0].childCount))
```
I'm not sure if this is something we want to braille. The actual suggestion is also displayed with a message, so the suggestion count could suppress that or vice versa. Keep in mind the default message timeout is 4 seconds, so a collision seems likely.
I think showing the index, e.g. 1 of 9, is enough. We already do that. In other words, I would use this message for speech only.

jcsteh commented:
@dkager, keep in mind that even though the first suggestion gets selected in the Start Menu (which is what most are familiar with), the first suggestion does not get selected elsewhere (e.g. searching in Settings). This PR is targeted at the latter case, wherein a suggestion would not be otherwise reported without user interaction (down arrow, etc.). Having said that, I still think this is potentially undesirable; that's being discussed in the issue at present.

josephsl commented:
Hi, @JCSTeh answered this. I’m holding this off until we get some things sorted out, such as use cases, performance issues and what not. Thanks.

 

From: Davy Kager [mailto:notifications@github.com] 
Sent: Wednesday, June 28, 2017 4:14 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA/SearchField: announce how many suggestions are available when entering text. re #7330 (#7331)

 

@dkager requested changes on this pull request.

  _____  

In source/NVDAObjects/UIA/__init__.py <https://github.com/nvaccess/nvda/pull/7331#discussion_r124513569> :

> @@ -1465,6 +1467,17 @@ def event_UIA_controllerFor(self):
               else:
                       self.event_suggestionsClosed()
 
+       def event_suggestionsOpened(self):
+              super(SearchField, self).event_suggestionsOpened()
+              # #7330: Announce number of items found (except in Start search box where the suggestions are selected as user types).
+              # Oddly, Edge's address omnibar returns 0 for suggestion count when there are clearly suggestions (implementation differences).
+              # Because inaccurate count could be announced (when users type, suggestion count changes), thus announce this if position info reporting is enabled.
+              if config.conf["presentation"]["reportObjectPositionInformation"]:
+                      if self.UIAElement.cachedAutomationID == "TextBox" or self.UIAElement.cachedAutomationID == "SearchTextBox" and self.appModule.appName != "searchui":
+                              # Item count must be the last one announced.
+                              # Translators: presented when there are suggestions as theu ser types (example output: 3 suggestions).
+                              queueHandler.queueFunction(queueHandler.eventQueue, ui.message, _("{suggestionsCount} suggestions").format(suggestionsCount = self.controllerFor[0].childCount))

I'm not sure if this is something we want to braille. The actual suggestion is also displayed with a message, so the suggestion count could suppress that or vice versa. Keep in mind the default message timeout is 4 seconds, so a collision seems likely.
I think showing the index, e.g. 1 of 9, is enough. We already do that. In other words, I would use this message for speech only.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/7331#pullrequestreview-46807662> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkNCQk41gaM1KZqKgVFupwgD1QyWmks5sIjWHgaJpZM4OHYuZ> .

Brian1Gaff commented:
I often feel that suggestions always changing as you type slugs things. One 
supposes that is why in google you can turn instant off.

Almost as annoying as auto complete was to some of us.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Joseph Lee" <notifications@github.com>
To: "nvaccess/nvda" <nvda@noreply.github.com>
Cc: "Subscribed" <subscribed@noreply.github.com>
Sent: Wednesday, June 28, 2017 4:45 PM
Subject: Re: [nvaccess/nvda] UIA/SearchField: announce how many suggestions 
are available when entering text. re #7330 (#7331)


Hi, @JCSTeh answered this. I’m holding this off until we get some things 
sorted out, such as use cases, performance issues and what not. Thanks.



From: Davy Kager [mailto:notifications@github.com]
Sent: Wednesday, June 28, 2017 4:14 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author 
<author@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA/SearchField: announce how many suggestions 
are available when entering text. re #7330 (#7331)



@dkager requested changes on this pull request.

  _____

In source/NVDAObjects/UIA/__init__.py 
<https://github.com/nvaccess/nvda/pull/7331#discussion_r124513569> :

> @@ -1465,6 +1467,17 @@ def event_UIA_controllerFor(self):
               else:
                       self.event_suggestionsClosed()

+       def event_suggestionsOpened(self):
+              super(SearchField, self).event_suggestionsOpened()
+              # #7330: Announce number of items found (except in Start 
search box where the suggestions are selected as user types).
+              # Oddly, Edge's address omnibar returns 0 for suggestion 
count when there are clearly suggestions (implementation differences).
+              # Because inaccurate count could be announced (when users 
type, suggestion count changes), thus announce this if position info 
reporting is enabled.
+              if 
config.conf["presentation"]["reportObjectPositionInformation"]:
+                      if self.UIAElement.cachedAutomationID == "TextBox" or 
self.UIAElement.cachedAutomationID == "SearchTextBox" and 
self.appModule.appName != "searchui":
+                              # Item count must be the last one announced.
+                              # Translators: presented when there are 
suggestions as theu ser types (example output: 3 suggestions).
+ 
queueHandler.queueFunction(queueHandler.eventQueue, ui.message, 
_("{suggestionsCount} suggestions").format(suggestionsCount = 
self.controllerFor[0].childCount))

I'm not sure if this is something we want to braille. The actual suggestion 
is also displayed with a message, so the suggestion count could suppress 
that or vice versa. Keep in mind the default message timeout is 4 seconds, 
so a collision seems likely.
I think showing the index, e.g. 1 of 9, is enough. We already do that. In 
other words, I would use this message for speech only.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub 
<https://github.com/nvaccess/nvda/pull/7331#pullrequestreview-46807662> , or 
mute the thread 
<https://github.com/notifications/unsubscribe-auth/AHgLkNCQk41gaM1KZqKgVFupwgD1QyWmks5sIjWHgaJpZM4OHYuZ> 
.



-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/nvaccess/nvda/pull/7331#issuecomment-311701382

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1465,6 +1467,17 @@ def event_UIA_controllerFor(self):
+		# Oddly, Edge's address omnibar returns 0 for suggestion count when there are clearly suggestions (implementation differences).
+		# Because inaccurate count could be announced (when users type, suggestion count changes), thus announce this if position info reporting is enabled.
+		if config.conf["presentation"]["reportObjectPositionInformation"]:
+			if self.UIAElement.cachedAutomationID == "TextBox" or self.UIAElement.cachedAutomationID == "SearchTextBox" and self.appModule.appName != "searchui":
```
Trivial, i'd write this as follows:
```
			if self.UIAElement.cachedAutomationID in ("TextBox","SearchTextBox") and self.appModule.appName != "searchui":
```

leonardder commented on user_docs/en/userGuide.t2t:
```diff
@@ -1133,6 +1133,7 @@ For example the File menu on a menu bar may have a shortcut key of alt+f.
 
 ==== Report object position information ====
 This option lets you choose whether you wish to have an object's position (e.g. 1 of 4) reported when moving to the object with the focus or object navigation.
+This is also used to announce number of suggestions available in some edit fields.
```
Looks somewhat vague to me. May be come up with at least one example?

josephsl commented:
Hi,

Membership test: no, I don't think it's a good idea to use that method. The resulting truth table is:

| TextBox | SearchTextBox | Not SearchUI | result |
| Yes | No | Yes | Item count is announced |
| Yes | No | No | Item count is not announced |
| No | Yes | Yes | Item count is announced |
| No | Yes | No | Item count is not announced |
| Yes | Yes | Yes | Item count is not announced |
| Yes | Yes | No | Implementation dependent |
| No | No | Yes | Undefined |
| No | No | No | Undefined |

This is simplified to: it'll check both automation ID's, as well as if the executable name is searchui or not.

But this is a general case that doesn't solve the heart of the problem: the heart of the problem, and the statement that makes a unique case for it, is the fact that there are controls with UIA ID of SearchTextBox that sometimes presents item count through moving the focus (Start menu/Cortana) and those that does not (Microsoft Store, for instance); fields marked with UIA ID of TextBox, on the other hand, falls into latter category: item count not announced automatically, and this was observed on ALL fields I can find. Thus the original statement (short-circuit via an early "or" statement) is stronger and simpler, and makes a more impactful solution to this issue. It also takes advantage of the fact that Python evaluates conditionals from left to right, and for "or" statement, it'll choose the correct execution path if the first statement alone is true. Perhaps I can make the original statement strong by putting the right hand side of the or statement inside parentheses.

Comments from others are appreciated. I'll take care of conflicts if and only if the PR is given a green light from @MichaelDCurran and others. Thanks.

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [16 okt. 2017 08:39 CEST](https://github.com/nvaccess/nvda/pull/7331#issuecomment-336793807 "2017-10-16T06:39:49Z - Replied by Github Reply Comments"):
> It also takes advantage of the fact that Python evaluates conditionals from left to right, and for "or" statement, it'll choose the correct execution path if the first statement alone is true. 

Ah yes, I missed that one.

> Perhaps I can make the original statement strong by putting the right hand side of the or statement inside parentheses.

I think that will improve readability a lot.

leonardder commented:
@josephsl: What's the status of this?

josephsl commented:
Hi, at some point I would like to see this come to NVDA Core, but it is up to users to say yes to this suggestion. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Tuesday, May 14, 2019 2:51 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA/SearchField: announce how many suggestions are available when entering text. re #7330 (#7331)

 

@josephsl <https://github.com/josephsl> : What's the status of this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/7331?email_source=notifications&email_token=AB4AXED54Y6PW2KNGJJCC53PVKDQZA5CNFSM4DQ5ROM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVK6QJI#issuecomment-492169253> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEGLZWKDA4KOS67BHC3PVKDQZANCNFSM4DQ5ROMQ> .

Adriani90 commented:
I think this would be very beneficial especially for braille display users. cc: @dkager, @Andre9642, @ruifontes. And it would also be useful for people who are not very familiar with screen readers, especially for new users.

josephsl commented:
Hi,

I may need to do a force push on this PR to make Appveyor happy...

Thanks.

josephsl committed:
Merged master

josephsl committed:
Search suggestions count: lint fixes.

Lint fixes include splitting long lines and spaces around str.format function call.

leonardder commented:
@josephsl This has been abandoned for a long time. There's also https://github.com/nvaccess/nvda/issues/7330#issuecomment-311538767 where @jcsteh expressed his concerns with the approach, which @Qchristensen and I share.

May be close this pr and bring this back to the design table in #7330?

josephsl commented:
Hi,

Nope, not abandoned - it's been a few weeks since last committing to it. However, since things have changed greatly since the initial commit, let's have a fresh discussion. Closing.
